include/diy/master.hpp

```diff
@@ -861,211 +870,307 @@ void
+                }
+            }
+        }                                       // for (OutgoingQueues::iterator it ...
+    }                                           // while (inflight_sends_.size() ...
+}
 
-        ir.message.buffer.reserve(msg_size);
-      }
-      else // tags::queue
-      {
-        diy::load_back(bb, ir.info);
-        ir.message.swap(bb);
-      }
-    }
-    else
+void
+diy::Master::
+check_incoming_queues(bool remote)
+{
+    mpi::optional<mpi::status> ostatus = comm_.iprobe(mpi::any_source, mpi::any_tag);
+    while (ostatus)
     {
-      size_t start_idx = ir.message.buffer.size();
-      size_t count = ostatus->count<char>();
-      ir.message.buffer.resize(start_idx + count);
+        InFlightRecv &ir = inflight_recvs_[ostatus->source()];
 
-      detail::VectorWindow<char> window;
-      window.begin = &ir.message.buffer[start_idx];
-      window.count = count;
-
-      comm_.recv(ostatus->source(), ostatus->tag(), window);
-    }
+        if (ir.info.from == -1) // uninitialized
+        {
+            MemoryBuffer bb;
+            comm_.recv(mpi::any_source, mpi::any_tag, bb.buffer);
+
+            if (ostatus->tag() == tags::piece)
+            {
+                size_t msg_size;
+                diy::load(bb, msg_size);
+                diy::load(bb, ir.info);
+
+                ir.message.buffer.reserve(msg_size);
+            }
+            else // tags::queue
+            {
+                diy::load_back(bb, ir.info);
+                ir.message.swap(bb);
+            }
+        }
+        else
+        {
+            size_t start_idx = ir.message.buffer.size();
+            size_t count = ostatus->count<char>();
+            ir.message.buffer.resize(start_idx + count);
 
-    if (ostatus->tag() == tags::queue)
-    {
-      size_t size  = ir.message.size();
-      int from = ir.info.from;
-      int to = ir.info.to;
-      int external = -1;
+            detail::VectorWindow<char> window;
+            window.begin = &ir.message.buffer[start_idx];
+            window.count = count;
 
-      assert(ir.info.round >= exchange_round_);
-      IncomingRound *in = &incoming_[ir.info.round];
+            comm_.recv(mpi::any_source, mpi::any_tag, window);
+        }
 
-      bool unload_queue = ((ir.info.round == exchange_round_) ? (block(lid(to)) == 0) : (limit_ != -1)) &&
-                          queue_policy_->unload_incoming(*this, from, to, size);
-      if (unload_queue)
-      {
-        log->debug("Directly unloading queue {} <- {}", to, from);
-        external = storage_->put(ir.message); // unload directly
-      }
-      else
-      {
-        in->map[to].queues[from].swap(ir.message);
-        in->map[to].queues[from].reset();     // buffer position = 0
-      }
-      in->map[to].records[from] = QueueRecord(size, external);
+        // unload message
 
-      ++(in->received);
-      ir = InFlightRecv(); // reset
-    }
+        if (remote || (!remote && ostatus->tag() == tags::queue))
```

Comment:
Why is `remote` treated differently here? If the queues are too large, they get split up using the same mechanism in both remote and ordinary setting. I think this should be just `if (ostatus->tag() == tags::queue)`.